ndb/src/google/cloud/ndb/model.py

```diff
@@ -2947,28 +2962,286 @@ def __init__(self, *args, **kwargs):
+                return kwargs.pop(keyword)
+
+        return None
+
+    def _set_attributes(self, kwargs):
+        """Set attributes from keyword arguments.
+
+        Args:
+            kwargs (Dict[str, Any]): A keyword arguments dictionary.
+        """
+        cls = type(self)
+        for name, value in kwargs.items():
+            # NOTE: This raises an ``AttributeError`` for unknown properties
+            #       and that is the intended behavior.
+            prop = getattr(cls, name)
+            if not isinstance(prop, Property):
+                raise TypeError("Cannot set non-property {}".format(name))
+            prop._set_value(self, value)
 
     @classmethod
     def _get_kind(cls):
-        """Return the kind name for this class.
+        """str: Return the kind name for this class.
 
         This defaults to ``cls.__name__``; users may override this to give a
         class a different name when stored in Google Cloud Datastore than the
         name of the class.
         """
         return cls.__name__
 
+    def _set_projection(self, projection):
+        """Set the projected properties for this instance.
+
+        Args:
+            projection (Union[list, tuple]): An iterable of strings
+                representing the projection for the model instance.
+        """
+        self._projection = tuple(projection)
+
+    @classmethod
+    def _fix_up_properties(cls):
+        """Fix up the properties by calling their ``_fix_up()`` method.
+
+        .. note::
+
+            This is called by :class:`MetaModel`, but may also be called
+            manually after dynamically updating a model class.
+
+        Raises:
+            KindError: If the returned kind from ``_get_kind()`` is not a
+                :class:`str`.
+            TypeError: If a property on this model has a name beginning with
+                an underscore.
+        """
+        kind = cls._get_kind()
+        if not isinstance(kind, str):
+            raise KindError(
+                "Class {} defines a ``_get_kind()`` method that returns "
+                "a non-string ({!r})".format(cls.__name__, kind)
+            )
+
+        # Map of ``kind_name`` -> ``property_instance``.
+        cls._properties = {}
+
+        # Skip the classes in ``ndb.model``.
+        if cls.__module__ == __name__:
+            return
+
+        for name in set(dir(cls)):
+            attr = getattr(cls, name, None)
+            if isinstance(attr, ModelAttribute) and not isinstance(
+                attr, ModelKey
+            ):
+                if name.startswith("_"):
+                    raise TypeError(
+                        "ModelAttribute {} cannot begin with an underscore "
+                        "character. ``_`` prefixed attributes are reserved "
+                        "for temporary Model instance values.".format(name)
+                    )
+                attr._fix_up(cls, name)
+                if isinstance(attr, Property):
```

Comment:
I am very unclear on why this branch is here. The `isinstance(attr, ModelAttribute)` check above seems to match some callers that are subclasses of `ModelAttribute`, but not `Property` subclasses. Such types don't seem to exist anywhere so I have no idea who / what is being supported by this.

Reply:
Maybe just an overabundance of caution? One solution might be to move what little is defined in ModelAttribute and move it to Property, if, as it seems, Property is the sole subclass of ModelAttribute.

Or leave it as-is. Either way.

Reply:
I think we (for now) have to leave it because `ModelAttribute` is public.

Reply:
Ok.